changefeedccl: Add a mechanism to throttle buffer drain rate.

miretskiy opened the pull request:
Make it possible to specify per-node changefeed configuration
`changefeed.node_sink_throttle_config` which enables throttling
the rate of drain of the memory buffer.

The throttling can be configure to limit the rate of message
emmission, the size of the emitted messages, as well as the
sink flush counts.  All of the throttles are specified as a
base value in units per second as well as burst rate:

```
{"MessageRate": 100, "MessageBurst": 200, "FlushRate": 1},
```

If the value is unspecified, the throttling is disabled.
The throttling configuration does not imply exact match on the
observed emitted values.  This is primarily due to per-sink
buffering behavior.  However, the throttling apply to the
buffer drain rate will have proportinoal impact on the
traffic emitted to the downstream sink.

When throttling is active, and the traffic is throttled, the changefeed
enters a "pushback" mode.  If the throttling configuration is such
that the traffic throttled below the incoming rate of messages,
the changefeed will deal with this situation by eventually restarting
the rangefeeds and performing a catchup scan.  Note, however, that such
regime is detrimental for overall performance of changefeeds.
Throttling should be used in order to limit short burst of traffic.

Follow on PRs will add throttling to the sinks themselves, as well
as add monitoring metrics.

Informs cockroachdb#58967

Release Notes: Changefeeds can be configured to throttle the rate
of emission from the memory buffer thus making it possible to limit
the emission rate from the changefeeds.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/67169)
<!-- Reviewable:end -->

Commit:
changefeedccl: Add a mechanism to throttle buffer drain rate.

Make it possible to specify per-node changefeed configuration
`changefeed.node_throttle_config` which enables throttling
the rate of drain of the memory buffer.

The throttling can be configure to limit the rate of message
emmission, the size of the emitted messages, as well as the
sink flush counts. All of the throttles are specified as a
base value in units per second as well as burst rate:

```
{"MessageRate": 100, "MessageBurst": 200, "FlushRate": 1}
```

If the value is unspecified, the throttling is disabled.
The throttling configuration does not imply exact match on the
observed emitted values. This is primarily due to per-sink
buffering behavior. However, the throttling apply to the
buffer drain rate will have proportinoal impact on the
traffic emitted to the downstream sink.

When throttling is active, and the traffic is throttled, the changefeed
enters a "pushback" mode. If the throttling configuration is such
that the traffic throttled below the incoming rate of messages,
the changefeed will deal with this situation by eventually restarting
the rangefeeds and performing a catchup scan. Note, however, that such
regime is detrimental for overall performance of changefeeds.
Throttling should be used in order to limit short burst of traffic.

Follow on PRs will add throttling to the sinks themselves, as well
as add monitoring metrics.

Informs #58967

Release Notes: Changefeeds can be configured to throttle the rate
of emission from the memory buffer thus making it possible to limit
the emission rate from the changefeeds.

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67169)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @HonoreDB, and @miretskiy)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/67169#-Mdvyrt97tRNRMC6Q8tG:-Mdvyrt97tRNRMC6Q8tH:bbifhim) (no related file):*
> This is primarily due to per-sink
> buffering behavior. 

Also the fact that the current implementation measures the bytes of the unencoded values and not the encoded values that might be on the wire.  I imagine for a format like JSON this can be a non-trivial under-estimate.

> Follow on PRs will add throttling to the sinks themselves, as well
as add monitoring metrics.

I think in this case we might want to think of a different name for the option itself.

---
___
*[pkg/ccl/changefeedccl/cdcutils/throttle.go, line 54 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67169#-MdvzcUm4sYe-0Ryjv99:-MdvzcUm4sYe-0Ryjv9A:bnr522x) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed1dea98dc3ee62dae1fdc8d3454538e5132802f/pkg/ccl/changefeedccl/cdcutils/throttle.go#L54)):*
> ```Go
> // AcquireFlushQuota acquires quota for a message with the specified size.
> // Blocks until such quota is available.
> func (t *Throttler) AcquireFlushQuota(ctx context.Context) error {
> ```

Looks like this potentially isn't wired into any calling code yet?
___
*[pkg/ccl/changefeedccl/kvevent/blocking_buffer.go, line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67169#-Mdw2qNB1iOFmJRdH2dV:-Mdw2qNB1iOFmJRdH2dW:b2leyl2) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed1dea98dc3ee62dae1fdc8d3454538e5132802f/pkg/ccl/changefeedccl/kvevent/blocking_buffer.go#L57)):*
> ```Go
> 		b.nodeThrottler = cdcutils.NodeLevelThrottler(sv)
> 	})
> }
> ```

I'm OK with the option, but I wonder if this might be more generally useful if we had a buffer that wrapped another buffer but enforced the quota.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67169)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @HonoreDB, and @stevendanna)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/67169#-Mdvyrt97tRNRMC6Q8tG:-Mdw4cU6701EHMkTauDQ:b-6cmfd) (no related file):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

> This is primarily due to per-sink
> buffering behavior. 

Also the fact that the current implementation measures the bytes of the unencoded values and not the encoded values that might be on the wire.  I imagine for a format like JSON this can be a non-trivial under-estimate.

> Follow on PRs will add throttling to the sinks themselves, as well
as add monitoring metrics.

I think in this case we might want to think of a different name for the option itself.

</blockquote></details>

Do you have suggestions?

---
___
*[pkg/ccl/changefeedccl/cdcutils/throttle.go, line 54 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67169#-MdvzcUm4sYe-0Ryjv99:-Mdw4elP0Ohl7aVhBFYa:bj99iv1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed1dea98dc3ee62dae1fdc8d3454538e5132802f/pkg/ccl/changefeedccl/cdcutils/throttle.go#L54)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Looks like this potentially isn't wired into any calling code yet?
</blockquote></details>

Yeah -- just the test.  I was debating if I should just drop it.  But pretty sure we'd want something like this for the sinks at some point.
___
*[pkg/ccl/changefeedccl/kvevent/blocking_buffer.go, line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67169#-Mdw2qNB1iOFmJRdH2dV:-Mdw4wIKDJLsMywxJfOF:br5boed) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed1dea98dc3ee62dae1fdc8d3454538e5132802f/pkg/ccl/changefeedccl/kvevent/blocking_buffer.go#L57)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

I'm OK with the option, but I wonder if this might be more generally useful if we had a buffer that wrapped another buffer but enforced the quota.
</blockquote></details>

Probably just as you typed it, I pushed an update with exactly that.


<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67169)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @HonoreDB, @miretskiy, and @stevendanna)

___
*[pkg/ccl/changefeedccl/kvevent/blocking_buffer.go, line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67169#-Mdw2qNB1iOFmJRdH2dV:-Mdw3uxQ8XWaRqGPvLAS:b56ozxj) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed1dea98dc3ee62dae1fdc8d3454538e5132802f/pkg/ccl/changefeedccl/kvevent/blocking_buffer.go#L57)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Probably just as you typed it, I pushed an update with exactly that.
</blockquote></details>

:D Looks good!


<!-- Sent from Reviewable.io -->

miretskiy commented:
@HonoreDB @stevendanna @ajwerner anybody has anymore comments on this?

amruss commented:
There are a couple of limitations here:
(1) You can't set at a per-changefeed level
(2) Throttling at this level means batching settings can still "override" the throttle downstream 

This is probably a really good first step as, but I'm guessing we'll still want to implement more per sink throttling

HonoreDB reviewed (commented):
> Reviewed 3 of 12 files at r1, 6 of 10 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67169)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy and @stevendanna)

___
*[docs/generated/settings/settings.html, line 7 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67169#-Me6B1kZ4lc-Uc_ESBtw:-Me6B1kZ4lc-Uc_ESBtx:b978ex8) ([raw file](https://github.com/cockroachdb/cockroach/blob/d416d5df805ae5bced09d45ad3f886cfb8d64511/docs/generated/settings/settings.html#L7)):*
> ```HTML
> <tr><td><code>admission.sql_kv_response.enabled</code></td><td>boolean</td><td><code>false</code></td><td>when true, work performed by the SQL layer when receiving a KV response is subject to admission control</td></tr>
> <tr><td><code>bulkio.stream_ingestion.minimum_flush_interval</code></td><td>duration</td><td><code>5s</code></td><td>the minimum timestamp between flushes; flushes may still occur if internal buffers fill up</td></tr>
> <tr><td><code>changefeed.node_sink_throttle_config</code></td><td>string</td><td><code></code></td><td>specifies node level throttling configuration for all changefeeeds</td></tr>
> ```

Suggest taking "sink" out of the config name as everything currently implemented is sink-agnostic.
___
*[pkg/ccl/changefeedccl/cdcutils/throttle.go, line 70 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67169#-Me6CM53Dob872EaGeJM:-Me6CM53Dob872EaGeJN:bu7tmxq) ([raw file](https://github.com/cockroachdb/cockroach/blob/d416d5df805ae5bced09d45ad3f886cfb8d64511/pkg/ccl/changefeedccl/cdcutils/throttle.go#L70)):*
> ```Go
> 	setLimits := func(rl *quotapool.RateLimiter, rate, burst float64) {
> 		// set rateBudget to unlimited if rate is 0.
> 		rateBudget := quotapool.Limit(math.MaxInt64)
> ```

Does quotapool short-circuit if it sees MaxInt64? If not, I'm worried we're doing unnecessary work somewhere when budget isn't set.
___
*[pkg/ccl/changefeedccl/kvevent/event.go, line 83 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67169#-Me6EoRrBeh4AbbLDiKt:-Me6EoRrBeh4AbbLDiKu:bmat0oh) ([raw file](https://github.com/cockroachdb/cockroach/blob/d416d5df805ae5bced09d45ad3f886cfb8d64511/pkg/ccl/changefeedccl/kvevent/event.go#L83)):*
> ```Go
> func (b *Event) ApproximateSize() int {
> 	if b.kv.Key != nil {
> 		return b.kv.Size() + b.prevVal.Size()
> ```

Can't prevVal be nil?


<!-- Sent from Reviewable.io -->

miretskiy commented:
@amruss Agree with all of your points.
Re `1) You can't set at a per-changefeed level` -- that wasn't the intention of this PR.
This builds out the infrastructure to implement per-changefeed level throttling. And it also gives a way to rate limit
all changefeeds.  As you say, per sink throttling should also be added.

miretskiy commented:
@HonoreDB @amruss Any additional comments/updates on this PR?

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67169)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB and @stevendanna)

___
*[docs/generated/settings/settings.html, line 7 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67169#-Me6B1kZ4lc-Uc_ESBtw:-Mf2ysW3CHswgk1XNeCH:bb6sh2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/d416d5df805ae5bced09d45ad3f886cfb8d64511/docs/generated/settings/settings.html#L7)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

Suggest taking "sink" out of the config name as everything currently implemented is sink-agnostic.
</blockquote></details>

Ack.
___
*[pkg/ccl/changefeedccl/cdcutils/throttle.go, line 54 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67169#-MdvzcUm4sYe-0Ryjv99:-Mf2yvCE53QplZ0-640K:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed1dea98dc3ee62dae1fdc8d3454538e5132802f/pkg/ccl/changefeedccl/cdcutils/throttle.go#L54)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Yeah -- just the test.  I was debating if I should just drop it.  But pretty sure we'd want something like this for the sinks at some point.
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/cdcutils/throttle.go, line 70 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67169#-Me6CM53Dob872EaGeJM:-Mf2zDQB2UtTzYzp50jw:b-1p4c97) ([raw file](https://github.com/cockroachdb/cockroach/blob/d416d5df805ae5bced09d45ad3f886cfb8d64511/pkg/ccl/changefeedccl/cdcutils/throttle.go#L70)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

Does quotapool short-circuit if it sees MaxInt64? If not, I'm worried we're doing unnecessary work somewhere when budget isn't set.
</blockquote></details>

It's not free, but it's basically the cost of a lock/unlock.  I wouldn't worry about this just yet -- we can always add "fast" path if profiles show this contributing to excessive utilization.
___
*[pkg/ccl/changefeedccl/kvevent/event.go, line 83 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67169#-Me6EoRrBeh4AbbLDiKt:-Mf2zQkv3q_EtLciLREU:b-bqkyzk) ([raw file](https://github.com/cockroachdb/cockroach/blob/d416d5df805ae5bced09d45ad3f886cfb8d64511/pkg/ccl/changefeedccl/kvevent/event.go#L83)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

Can't prevVal be nil?
</blockquote></details>

No, it cannot.  prevVal is a roachpb.Value protocol buffer (not a pointer).  It may as well have "nil" data, but size method on the non-nil protocol message works correctly.
___
*[pkg/ccl/changefeedccl/kvevent/blocking_buffer.go, line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67169#-Mdw2qNB1iOFmJRdH2dV:-Mf2z_rDAw6yhbQ79CCM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed1dea98dc3ee62dae1fdc8d3454538e5132802f/pkg/ccl/changefeedccl/kvevent/blocking_buffer.go#L57)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

:D Looks good!
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

HonoreDB reviewed (approved):
No more comments. I think the build failure is you need to regenerate docs.

> Reviewed 1 of 12 files at r1, 1 of 10 files at r3, 7 of 7 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67169#-:-Mf3I7nF9AdaAKOQS_5F:bnjf1h6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB and @stevendanna)



<!-- Sent from Reviewable.io -->

amruss commented:
LGTM

miretskiy commented:
tftr
bors r+
